File: app/src/main/kotlin/com/saatxi/eatapp/ui/list/RestaurantListScreen.kt
package com.saatxi.eatapp.ui.list

import android.content.res.Configuration
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Add
import androidx.compose.material.icons.filled.Share
import androidx.compose.material.icons.outlined.RestaurantMenu
import androidx.compose.material.icons.outlined.SearchOff
import androidx.compose.material3.Button
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.FloatingActionButton
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.material3.TopAppBar
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.pluralStringResource
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.lifecycle.viewmodel.compose.viewModel
import com.saatxi.eatapp.R
import com.saatxi.eatapp.ui.AppViewModelProvider
import com.saatxi.eatapp.ui.common.DeleteConfirmDialog
import com.saatxi.eatapp.ui.model.RestaurantUiModel
import com.saatxi.eatapp.ui.theme.EatAppTheme

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun RestaurantListScreen(
    onOpenRestaurant: (Long) -> Unit,
    onAddRestaurant: () -> Unit,
    viewModel: RestaurantListViewModel = viewModel(factory = AppViewModelProvider.Factory)
) {
    val uiState by viewModel.uiState.collectAsState()
    val context = LocalContext.current
    // Set by a row's swipe-to-delete gesture — see RestaurantRow's own
    // onDeleteRequest doc for why the row itself never deletes directly.
    var pendingDelete by remember { mutableStateOf<RestaurantUiModel?>(null) }

    pendingDelete?.let { restaurant ->
        DeleteConfirmDialog(
            onConfirm = {
                viewModel.onDeleteRestaurant(restaurant.id)
                pendingDelete = null
            },
            onDismiss = { pendingDelete = null }
        )
    }

    Scaffold(
        topBar = {
            Column {
                TopAppBar(
                    title = { Text(stringResource(R.string.list_title)) },
                    actions = {
                        IconButton(onClick = { viewModel.onShareAll(context) }) {
                            Icon(Icons.Default.Share, contentDescription = stringResource(R.string.list_action_share_all))
                        }
                    }
                )
                HorizontalDivider(color = MaterialTheme.colorScheme.outlineVariant)
            }
        },
        floatingActionButton = {
            FloatingActionButton(onClick = onAddRestaurant) {
                Icon(Icons.Default.Add, contentDescription = stringResource(R.string.list_action_add_restaurant))
            }
        }
    ) { padding ->
        Column(modifier = Modifier.fillMaxSize().padding(padding)) {
            SearchAndFilterBar(
                searchQuery = uiState.searchQuery,
                onSearchQueryChange = viewModel::onSearchQueryChange,
                // Same condition the list content below switches its empty state
                // on: nothing to sort or filter yet during the first load, or
                // before any restaurant has ever been added.
                showSortAndFilters = !uiState.isInitialLoad && (uiState.restaurants.isNotEmpty() || uiState.hasActiveFilter),
                sort = uiState.sort,
                onSortChange = viewModel::onSortChange,
                minRating = uiState.minRating,
                onMinRatingChange = viewModel::onMinRatingChange,
                cuisineType = uiState.cuisineType,
                availableCuisines = uiState.availableCuisines,
                onCuisineChange = viewModel::onCuisineChange,
                visited = uiState.visited,
                onVisitedChange = viewModel::onVisitedChange
            )

            Box(modifier = Modifier.weight(1f).fillMaxWidth()) {
                if (uiState.isInitialLoad) {
                    // The database has not emitted yet, so an empty list here means
                    // "not loaded", not "nothing to show" — painting the empty state
                    // would flash it for a frame on every cold start. Shape-matching
                    // skeleton rows (F-67) read as faster than a centred spinner even
                    // though the actual wait is identical.
                    Column(
                        modifier = Modifier.padding(16.dp),
                        verticalArrangement = Arrangement.spacedBy(8.dp)
                    ) {
                        repeat(SKELETON_ROW_COUNT) { RestaurantRowSkeleton() }
                    }
                } else if (uiState.restaurants.isEmpty() && !uiState.hasActiveFilter) {
                    // Nothing has ever been added: there are no filters to offer yet.
                    EmptyState(
                        icon = Icons.Outlined.RestaurantMenu,
                        title = stringResource(R.string.list_empty_title),
                        body = stringResource(R.string.list_empty_body),
                        modifier = Modifier.fillMaxSize(),
                        actionLabel = stringResource(R.string.list_action_add_restaurant),
                        onAction = onAddRestaurant
                    )
                } else {
                    LazyColumn(
                        contentPadding = PaddingValues(16.dp),
                        verticalArrangement = Arrangement.spacedBy(8.dp)
                    ) {
                        if (uiState.restaurants.isEmpty()) {
                            item(key = "no-results", contentType = "empty") {
                                EmptyState(
                                    icon = Icons.Outlined.SearchOff,
                                    title = stringResource(R.string.list_empty_no_results_title),
                                    body = stringResource(R.string.list_empty_no_results_body),
                                    actionLabel = stringResource(R.string.list_action_clear_filters),
                                    onAction = viewModel::clearFilters,
                                    modifier = Modifier.fillParentMaxHeight(0.6f).animateItem()
                                )
                            }
                        } else {
                            // Most useful precisely when a filter has narrowed the list down —
                            // an unfiltered count adds nothing you can't already see. When
                            // browsing everything instead (F-66), that same spot offers a
                            // starting point rather than sitting blank.
                            if (uiState.hasActiveFilter) {
                                item(key = "result-count", contentType = "header") {
                                    Text(
                                        text = pluralStringResource(
                                            R.plurals.list_result_count,
                                            uiState.restaurants.size,
                                            uiState.restaurants.size
                                        ),
                                        style = MaterialTheme.typography.labelMedium,
                                        color = MaterialTheme.colorScheme.onSurfaceVariant,
                                        modifier = Modifier.animateItem()
                                    )
                                }
                            } else {
                                item(key = "suggestions", contentType = "header") {
                                    val topCuisine = remember(uiState.restaurants) {
                                        uiState.restaurants.groupingBy { it.cuisineKey }.eachCount().maxByOrNull { it.value }?.key
                                    }
                                    SearchSuggestionsRow(
                                        topCuisine = topCuisine,
                                        onMinRatingChange = viewModel::onMinRatingChange,
                                        onVisitedChange = viewModel::onVisitedChange,
                                        onCuisineChange = viewModel::onCuisineChange,
                                        modifier = Modifier.animateItem()
                                    )
                                }
                            }
                            items(uiState.restaurants, key = { it.id }, contentType = { "restaurant" }) { restaurant ->
                                RestaurantRow(
                                    restaurant = restaurant,
                                    onClick = { onOpenRestaurant(restaurant.id) },
                                    onFavoriteToggle = viewModel::onFavoriteToggle,
                                    onDeleteRequest = { pendingDelete = restaurant },
                                    modifier = Modifier.animateItem()
                                )
                            }
                        }
                    }
                }
            }
        }
    }
}

/** Internal rather than private: reused by [com.saatxi.eatapp.ui.favorites.FavoritesScreen]. */
@Composable
internal fun EmptyState(
    icon: ImageVector,
    title: String,
    body: String,
    modifier: Modifier = Modifier,
    actionLabel: String? = null,
    onAction: () -> Unit = {},
    actionEnabled: Boolean = true
) {
    Box(modifier = modifier, contentAlignment = Alignment.Center) {
        Column(
            horizontalAlignment = Alignment.CenterHorizontally,
            modifier = Modifier.padding(32.dp)
        ) {
            Icon(
                icon,
                contentDescription = null,
                modifier = Modifier.size(64.dp),
                tint = MaterialTheme.colorScheme.onSurfaceVariant
            )
            Text(
                text = title,
                style = MaterialTheme.typography.titleMedium,
                modifier = Modifier.padding(top = 16.dp)
            )
            Text(
                text = body,
                style = MaterialTheme.typography.bodyLarge,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
                modifier = Modifier.padding(top = 4.dp)
            )
            if (actionLabel != null) {
                Button(onClick = onAction, enabled = actionEnabled, modifier = Modifier.padding(top = 20.dp)) {
                    Text(actionLabel)
                }
            }
        }
    }
}

@Preview(name = "Light")
@Preview(name = "Dark", uiMode = Configuration.UI_MODE_NIGHT_YES)
@Composable
private fun EmptyStateFirstAddPreview() {
    EatAppTheme {
        Surface {
            EmptyState(
                icon = Icons.Outlined.RestaurantMenu,
                title = stringResource(R.string.list_empty_title),
                body = stringResource(R.string.list_empty_body),
                modifier = Modifier.fillMaxSize(),
                actionLabel = stringResource(R.string.list_action_add_restaurant),
                onAction = {}
            )
        }
    }
}

@Preview(name = "Light")
@Preview(name = "Dark", uiMode = Configuration.UI_MODE_NIGHT_YES)
@Composable
private fun EmptyStateNoResultsPreview() {
    EatAppTheme {
        Surface {
            EmptyState(
                icon = Icons.Outlined.SearchOff,
                title = stringResource(R.string.list_empty_no_results_title),
                body = stringResource(R.string.list_empty_no_results_body),
                modifier = Modifier.fillMaxSize(),
                actionLabel = stringResource(R.string.list_action_clear_filters),
                onAction = {}
            )
        }
    }
}

File: lib/features/list/restaurant_list_screen.dart
import 'package:flutter/material.dart';

import '../../app/app_scope.dart';
import '../../core/l10n/generated/app_localizations.dart';
import '../../core/theme/tokens/app_spacing.dart';
import '../../core/widgets/delete_confirm_dialog.dart';
import '../../core/widgets/empty_state.dart';
import '../../core/widgets/staggered_entrance.dart';
import '../import_export/share_service.dart';
import 'restaurant_list_controller.dart';
import 'restaurant_row.dart';
import 'restaurant_ui_model.dart';
import 'search_and_filter_bar.dart';

/// The list of restaurants: search, sort and filters above, the rows below, and
/// the loading, empty and no-results states in between.
///
/// The screen owns its [RestaurantListController] and rebuilds from it through a
/// `ListenableBuilder`, which is the Flutter counterpart of the Android screen's
/// `collectAsState()` over its ViewModel's `StateFlow`. The controller is built
/// in [didChangeDependencies] rather than `initState` because it reads the
/// repositories off [AppScope], and an inherited-widget lookup is only legal
/// once dependencies have been established.
class RestaurantListScreen extends StatefulWidget {
  const RestaurantListScreen({
    super.key,
    this.onOpenRestaurant,
    this.onAddRestaurant,
    this.favouritesOnly = false,
  });

  /// Null leaves the rows untappable — the case in a bare widget test.
  final ValueChanged<RestaurantUiModel>? onOpenRestaurant;
  final VoidCallback? onAddRestaurant;

  /// Narrows the screen to favourites, which is what the favourites tab shows:
  /// the same search, sort, filters and rows, cut down to the hearted places and
  /// with nothing to add or suggest.
  final bool favouritesOnly;

  @override
  State<RestaurantListScreen> createState() => _RestaurantListScreenState();
}

class _RestaurantListScreenState extends State<RestaurantListScreen> {
  RestaurantListController? _controller;

  /// Owned here rather than by the bar so the screen can push the active query
  /// back into it when a filter change resets it.
  final TextEditingController _searchController = TextEditingController();

  @override
  void didChangeDependencies() {
    super.didChangeDependencies();
    final AppScope scope = AppScope.of(context);
    _controller ??= RestaurantListController(
      repository: scope.restaurants,
      preferences: scope.preferences,
      favouritesOnly: widget.favouritesOnly,
    );
  }

  @override
  void dispose() {
    _searchController.dispose();
    _controller?.dispose();
    super.dispose();
  }

  /// A row only ever requests a delete; the confirmation is shown here, and the
  /// row is removed only once it is accepted.
  Future<void> _confirmDelete(RestaurantUiModel restaurant) async {
    final bool confirmed = await showDeleteConfirmDialog(context);
    if (!confirmed) {
      return;
    }
    await _controller?.deleteRestaurant(restaurant.id);
  }

  @override
  Widget build(BuildContext context) {
    final RestaurantListController controller = _controller!;
    final AppLocalizations l10n = AppLocalizations.of(context);

    return Scaffold(
      appBar: AppBar(
        title: Text(
          widget.favouritesOnly ? l10n.favoritesTitle : l10n.listTitle,
        ),
        // "Share all" means all, not just what the active filters leave visible
        // — the same rule the Android screen follows. Favourites has nothing of
        // its own to export, so it drops the action.
        actions: widget.favouritesOnly
            ? null
            : <Widget>[
                IconButton(
                  onPressed: () => exportAndShareRestaurants(
                    context,
                    repository: AppScope.of(context).restaurants,
                  ),
                  tooltip: l10n.listActionShareAll,
                  icon: const Icon(Icons.share_outlined),
                ),
              ],
        // A two-stop tonal wash rather than a flat container colour: a sense of
        // place above the list without touching the app bar's scroll behaviour.
        flexibleSpace:
            widget.favouritesOnly ? null : const _TopGradient(),
      ),
      floatingActionButton: widget.onAddRestaurant == null
          ? null
          : FloatingActionButton(
              onPressed: widget.onAddRestaurant,
              tooltip: l10n.listActionAddRestaurant,
              child: const Icon(Icons.add),
            ),
      body: ListenableBuilder(
        listenable: controller,
        builder: (BuildContext context, Widget? child) {
          final RestaurantListUiState state = controller.state;
          // Keep the field in step with the state: clearing the filters resets
          // the query, and the field has to go blank with it.
          if (_searchController.text != state.searchQuery) {
            _searchController.value = TextEditingValue(
              text: state.searchQuery,
              selection: TextSelection.collapsed(
                offset: state.searchQuery.length,
              ),
            );
          }
          return Column(
            children: <Widget>[
              SearchAndFilterBar(
                searchController: _searchController,
                onSearchQueryChange: controller.onSearchQueryChange,
                // Nothing to sort or filter yet during the first load, or before
                // any restaurant has ever been added.
                showSortAndFilters: !state.isInitialLoad &&
                    (state.restaurants.isNotEmpty || state.hasActiveFilter),
                sort: state.sort,
                onSortChange: controller.onSortChange,
                minRating: state.minRating,
                onMinRatingChange: controller.onMinRatingChange,
                cuisineType: state.cuisineType,
                availableCuisines: state.availableCuisines,
                onCuisineChange: controller.onCuisineChange,
                visited: state.visited,
                onVisitedChange: controller.onVisitedChange,
                city: state.city,
                availableCities: state.availableCities,
                onCityChange: controller.onCityChange,
                region: state.region,
                availableRegions: state.availableRegions,
                onRegionChange: controller.onRegionChange,
                country: state.country,
                availableCountries: state.availableCountries,
                onCountryChange: controller.onCountryChange,
                priceRange: state.priceRange,
                onPriceRangeChange: controller.onPriceRangeChange,
              ),
              Expanded(
                child: RefreshIndicator(
                  // Pulling re-runs the query rather than waiting on a cosmetic
                  // delay — see RestaurantListController.refresh.
                  onRefresh: controller.refresh,
                  child: _content(state, controller, l10n),
                ),
              ),
            ],
          );
        },
      ),
    );
  }

  Widget _content(
    RestaurantListUiState state,
    RestaurantListController controller,
    AppLocalizations l10n,
  ) {
    if (state.isInitialLoad) {
      // The database has not emitted yet, so an empty list here means "not
      // loaded", not "nothing to show" — painting the empty state would flash it
      // for a frame on every cold start. Shape-matching skeleton rows read as
      // faster than a centred spinner even though the wait is identical.
      return ListView.separated(
        // Scrollable even when the content is shorter than the viewport, so the
        // pull-to-refresh gesture is available in every state.
        physics: const AlwaysScrollableScrollPhysics(),
        padding: const EdgeInsets.all(AppSpacing.lg),
        itemCount: skeletonRowCount,
        separatorBuilder: (BuildContext context, int index) =>
            const SizedBox(height: AppSpacing.sm),
        itemBuilder: (BuildContext context, int index) =>
            const RestaurantRowSkeleton(),
      );
    }

    final bool noResults = state.restaurants.isEmpty;

    if (noResults && !state.hasActiveFilter) {
      // There is nothing to narrow down yet — either no restaurant has been
      // added, or none has been hearted.
      return _refreshableEmpty(
        EmptyState(
          icon: widget.favouritesOnly
              ? Icons.favorite_border
              : Icons.restaurant_menu,
          title: widget.favouritesOnly
              ? l10n.favoritesEmptyTitle
              : l10n.listEmptyTitle,
          body: widget.favouritesOnly
              ? l10n.favoritesEmptyBody
              : l10n.listEmptyBody,
          actionLabel:
              !widget.favouritesOnly && widget.onAddRestaurant != null
                  ? l10n.listActionAddRestaurant
                  : null,
          onAction: widget.onAddRestaurant,
        ),
      );
    }

    // A result count is most useful precisely when a filter has narrowed the
    // list down; browsing everything instead, that same spot offers a starting
    // point rather than sitting blank. Favourites has nothing to suggest.
    final Widget? header = state.hasActiveFilter
        ? Align(
            alignment: Alignment.centerLeft,
            child: Text(
              l10n.listResultCount(state.restaurants.length),
              style: Theme.of(context).textTheme.labelMedium?.copyWith(
                color: Theme.of(context).colorScheme.onSurfaceVariant,
              ),
            ),
          )
        : widget.favouritesOnly
            ? null
            : SearchSuggestionsRow(
                onMinRatingChange: controller.onMinRatingChange,
                onVisitedChange: controller.onVisitedChange,
              );
    final int headerCount = header == null ? 0 : 1;

    return ListView.separated(
      physics: const AlwaysScrollableScrollPhysics(),
      padding: const EdgeInsets.all(AppSpacing.lg),
      itemCount: headerCount + (noResults ? 1 : state.restaurants.length),
      separatorBuilder: (BuildContext context, int index) =>
          const SizedBox(height: AppSpacing.sm),
      itemBuilder: (BuildContext context, int index) {
        if (header != null && index == 0) {
          return header;
        }
        if (noResults) {
          return SizedBox(
            height: 360,
            child: EmptyState(
              icon: Icons.search_off,
              title: l10n.listEmptyNoResultsTitle,
              body: l10n.listEmptyNoResultsBody,
              actionLabel: l10n.listActionClearFilters,
              onAction: controller.clearFilters,
            ),
          );
        }
        final RestaurantUiModel restaurant = state.restaurants[index - headerCount];
        return StaggeredEntrance(
          // Indexed by the row's place in the list, so the cascade runs top to
          // bottom whether or not the suggestion header is showing above it.
          index: index,
          child: RestaurantRow(
            restaurant: restaurant,
            onTap: widget.onOpenRestaurant == null
                ? null
                : () => widget.onOpenRestaurant!(restaurant),
            onFavoriteToggle: controller.toggleFavorite,
            onDeleteRequest: () => _confirmDelete(restaurant),
          ),
        );
      },
    );
  }

  /// Wraps a shorter-than-the-viewport state so it can still be pulled.
  ///
  /// A [RefreshIndicator] needs a scrollable child, and the empty states are
  /// centred blocks lighter than the screen — handed over raw, they would give
  /// the gesture nothing to grab.
  Widget _refreshableEmpty(Widget child) => LayoutBuilder(
    builder: (BuildContext context, BoxConstraints constraints) =>
        SingleChildScrollView(
          physics: const AlwaysScrollableScrollPhysics(),
          child: ConstrainedBox(
            constraints: BoxConstraints(minHeight: constraints.maxHeight),
            child: child,
          ),
        ),
  );
}

class _TopGradient extends StatelessWidget {
  const _TopGradient();

  @override
  Widget build(BuildContext context) {
    final ColorScheme scheme = Theme.of(context).colorScheme;
    return DecoratedBox(
      decoration: BoxDecoration(
        gradient: LinearGradient(
          begin: Alignment.topCenter,
          end: Alignment.bottomCenter,
          colors: <Color>[scheme.primaryContainer, scheme.surface],
        ),
      ),
    );
  }
}
